dagshub/mlflow/patch.py

```diff
@@ -0,0 +1,176 @@
+            if attr_name.startswith("_"):
+                continue
+
+            full_attr_name = f"{module.__name__}.{attr_name}"
+            attr = getattr(module, attr_name)
+
+            is_function = isinstance(attr, FunctionType)
+            is_class = inspect.isclass(attr)
+
+            if is_function:
+                if self.is_function_already_patched(full_attr_name):
+                    continue
+                self.original_func_lookup[full_attr_name] = attr
+                setattr(module, attr_name, self._wrap_func(attr))
+                continue
+
+            if is_class and full_attr_name in self.classes_to_patch:
+                for class_attr_name in dir(attr):
+                    if class_attr_name.startswith("_"):
+                        continue
+                    class_attr = getattr(attr, class_attr_name)
+                    full_class_attr_name = f"{full_attr_name}.{class_attr_name}"
+                    if isinstance(class_attr, FunctionType) and not self.is_function_already_patched(
+                        full_class_attr_name
+                    ):
+                        self.original_func_lookup[full_class_attr_name] = class_attr
+                        setattr(attr, class_attr_name, self._wrap_func(class_attr))
+
+            is_module = isinstance(attr, ModuleType)
+            # Ignore lazy loaded modules, don't want to load them all at once
+            # TODO: maybe let users tune this somehow, although from what I've seen most of those are wrappers,
+            #  and the actual mlflow logging functions are eagerly loaded
+            is_lazy_module = isinstance(attr, LazyLoader)
+
+            if not is_module or is_lazy_module:
+                continue
+
+            full_module_name = attr.__name__
+
+            if not full_module_name.startswith("mlflow"):
+                continue
+
+            if full_module_name in self.patched_modules:
+                continue
+
+            # Recurse into the module and patch it
+            self.monkey_patch_module(attr)
+
+    def unpatch(self):
+        logger.warning("Unpatching MLflow to raise exceptions again")
+        global _global_mlflow_patch
+
+        for unpatched_key, unpatched_func in self.original_func_lookup.items():
+            attrs = unpatched_key.split(".")
+            module_name = attrs[0]
+            attrs = attrs[1:]
+            attr = sys.modules[module_name]
+            for index, attr_name in enumerate(attrs):
+                parent_attr = attr
+                attr = getattr(parent_attr, attr_name)
+                if index == len(attrs) - 1:
+                    setattr(parent_attr, attr_name, unpatched_func)
+
+        self.patched_modules.clear()
+        self.original_func_lookup.clear()
+        _global_mlflow_patch = None
+
+
+_global_mlflow_patch: Optional[MlflowMonkeyPatch] = None
+
+
+def patch_mlflow():
```

Comment:
IMO, there should be a low-level global function that accepts the explicit lists of functions and classes to patch, and a higher level abstract function that accepts user-facing common flags:
1. `patch_artifact_uploads=True`
2. `additional_functions=[]`
3. Anything other high-level that you might think of. e.g. a default-off flag to patch `start_run`

Also, it would be extra nice to call the global patch function from our `dagshub.init(mlflow=True)` function, you can add an additional flag `patch_mlflow=False` by default